backend/de.metas.monitoring/src/main/java/de/metas/monitoring/adapter/MicrometerPerformanceMonitoringService.java

```diff
@@ -42,63 +42,88 @@
 public class MicrometerPerformanceMonitoringService implements PerformanceMonitoringService
 {
 	private final MeterRegistry meterRegistry;
-	private final Optional<APMPerformanceMonitoringService> apmPerformanceMonitoringService;
+	private static final ThreadLocal<PerformanceMonitoringData> perfMonDataTL = ThreadLocal.withInitial(PerformanceMonitoringData::new);
 
 	public MicrometerPerformanceMonitoringService(
-			@NonNull final Optional<APMPerformanceMonitoringService> apmPerformanceMonitoringService,
 			@NonNull final MeterRegistry meterRegistry)
 	{
-		this.apmPerformanceMonitoringService = apmPerformanceMonitoringService;
 		this.meterRegistry = meterRegistry;
 	}
 
 	@Override
-	public <V> V monitorTransaction(
+	public <V> V monitor(
 			@NonNull final Callable<V> callable,
-			@NonNull final TransactionMetadata metadata)
+			@NonNull final PerformanceMonitoringService.Metadata metadata)
 	{
-		final ArrayList<Tag> tags = createTagsFromLabels(metadata.getLabels());
-		mkTagIfNotNull("Name", metadata.getName()).ifPresent(tags::add);
 
-		final Callable<V> callableToUse;
-		if (apmPerformanceMonitoringService.isPresent())
+		final PerformanceMonitoringData perfMonData = perfMonDataTL.get();
+		final ArrayList<Tag> tags = new ArrayList<>();
+
+		mkTagIfNotNull("name", metadata.getClassName()).ifPresent(tags::add);
+		mkTagIfNotNull("action", metadata.getFunctionName()).ifPresent(tags::add);
+
+		if(perfMonData.getDepth() == 0
+				&& metadata.getType() == Type.REST_CONTROLLER_WITH_WINDOW_ID
+				&& metadata.getFunctionName() != null
+				&& metadata.getWindowNameAndId() != null)
 		{
-			callableToUse = () -> apmPerformanceMonitoringService.get().monitorTransaction(callable, metadata);
+			perfMonData.setIsInitiatorLabelActive(true);
+			perfMonData.setInitiator(metadata.getClassName() + " - " + metadata.getFunctionName());
+			perfMonData.setInitiatorWindow(metadata.getWindowNameAndId());
+			perfMonData.getCalledBy().add(0, "HTTP Request");
+
 		}
-		else
+		else if(perfMonData.getDepth() == 0
+				&& metadata.getType() == Type.REST_CONTROLLER
+				&& metadata.getFunctionName() != null)
 		{
-			callableToUse = callable;
+			perfMonData.setIsInitiatorLabelActive(true);
+			perfMonData.setInitiator(metadata.getClassName() + " - " + metadata.getFunctionName());
+			perfMonData.setInitiatorWindow("NONE");
+			perfMonData.getCalledBy().add(0, "HTTP Request");
```

Comment:
> perfMonData.getCalledBy().add(0, "HTTP Request");

in perfMonData class,
introduce a method called something like `addCalledByAsFirst`.

Btw, why are u inserting it at first position?
If depth is ZERO isn't it supposed to be the first call?

Reply:
`calledBy` label for depth=0 is needed for the filter inside the dashboard. So every function can find the label-value at it's own depth and not depth-1 and sets the value at depth+1.